File: functions/test/searchkey.test.js
const test = require("node:test");
const assert = require("node:assert");
const { normalize } = require("../lib/categories");

/**
 * The table below is duplicated verbatim in
 * app/src/test/java/com/safebeauty/app/util/SearchKeyTest.kt.
 *
 * It has to be. Firestore matches a prefix against the STORED value, so the
 * device must derive the same string the server wrote or the query is a range
 * over spelling that never existed. The server stored a salon as
 * "سالن ارایشی عروس خانم" while the app sent "سالن آرایشی…", and a customer
 * typing the name exactly as it appeared on her screen was told there were no
 * results.
 */
const CASES = [
  ["سالن آرایشی عروس خانم", "سالن ارایشی عروس خانم"],   // madda folded
  ["آرایشگاه زیبایی",       "ارایشگاه زیبایی"],
  ["نوري",                  "نوری"],                     // Arabic yeh → Persian
  ["حکيم الله",             "حکیم الله"],
  ["مکياژ",                 "مکیاژ"],
  ["زيبايي كابل",           "زیبایی کابل"],              // Arabic yeh AND kaf
  ["Shaghayeq Ha",          "shaghayeq ha"],
  ["  spaced   out  ",      "spaced out"],
  ["فاطمة",                 "فاطمه"],                    // teh marbuta → heh
  ["",                      ""],
];

test("normalize folds the spellings Dari and Pashto keyboards actually produce", () => {
  for (const [input, expected] of CASES) {
    assert.equal(normalize(input), expected, `normalize(${JSON.stringify(input)})`);
  }
});

test("normalizing twice changes nothing", () => {
  // The device normalizes a search term; the server normalized the stored name.
  // If the operation were not idempotent the two could still disagree.
  for (const [input] of CASES) {
    assert.equal(normalize(normalize(input)), normalize(input));
  }
});

test("a name and the same name typed on the other keyboard collapse together", () => {
  // The whole point: two spellings of one salon must not be two salons.
  assert.equal(normalize("زيبايي"), normalize("زیبایی"));
  assert.equal(normalize("كابل"), normalize("کابل"));
  assert.equal(normalize("آرایش"), normalize("ارایش"));
});

test("null and undefined are the empty key, not a crash", () => {
  assert.equal(normalize(null), "");
  assert.equal(normalize(undefined), "");
});

File: app/src/test/java/com/safebeauty/app/util/SearchKeyTest.kt
package com.safebeauty.app.util

import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * The same table as functions/test/searchkey.test.js, deliberately.
 *
 * Firestore matches a prefix against the stored value, so the device has to
 * derive exactly the string the server wrote. When it did not, a customer typing
 * a salon's name as it appeared on her screen got nothing back — which reads as
 * "this salon is not on SafeBeauty", not as a bug.
 */
class SearchKeyTest {

    private val cases = listOf(
        "سالن آرایشی عروس خانم" to "سالن ارایشی عروس خانم",
        "آرایشگاه زیبایی"       to "ارایشگاه زیبایی",
        "نوري"                  to "نوری",
        "حکيم الله"             to "حکیم الله",
        "مکياژ"                 to "مکیاژ",
        "زيبايي كابل"           to "زیبایی کابل",
        "Shaghayeq Ha"          to "shaghayeq ha",
        "  spaced   out  "      to "spaced out",
        "فاطمة"                 to "فاطمه",
        ""                      to "",
    )

    @Test
    fun `folds the spellings Dari and Pashto keyboards actually produce`() {
        for ((input, expected) in cases) {
            assertEquals("normalize(\"$input\")", expected, SearchKey.normalize(input))
        }
    }

    @Test
    fun `normalizing twice changes nothing`() {
        for ((input, _) in cases) {
            val once = SearchKey.normalize(input)
            assertEquals(once, SearchKey.normalize(once))
        }
    }

    @Test
    fun `two keyboards, one salon`() {
        assertEquals(SearchKey.normalize("زيبايي"), SearchKey.normalize("زیبایی"))
        assertEquals(SearchKey.normalize("كابل"), SearchKey.normalize("کابل"))
        assertEquals(SearchKey.normalize("آرایش"), SearchKey.normalize("ارایش"))
    }

    @Test
    fun `null is the empty key, not a crash`() {
        assertEquals("", SearchKey.normalize(null))
    }
}
